core/src/main/java/tech/ydb/core/impl/operation/OperationManager.java

```diff
@@ -0,0 +1,194 @@
+                                        );
+                                    }
+                                }
+                            }
+                    );
+                },
+                OPERATION_CHECK_TIMEOUT_MS,
+                TimeUnit.MILLISECONDS
+        );
+    }
+
+    private CompletableFuture<Result<OperationProtos.CancelOperationResponse>> cancel(
+            final Operation<?> operation
+    ) {
+        return transport.unaryCall(
+                OperationServiceGrpc.getCancelOperationMethod(),
+                GrpcRequestSettings.newBuilder()
+                        .build(),
+                OperationProtos.CancelOperationRequest.newBuilder()
+                        .setId(operation.operationId)
+                        .build()
+        ).whenComplete(
+                (cancelOperationResponseResult, throwable) -> {
+                    if (throwable != null) {
+                        logger.error("Fail cancel polling operation with id: {}", operation.operationId, throwable);
+                    }
+
+                    if (cancelOperationResponseResult.isSuccess()) {
+                        logger.info("Success cancel polling operation with id: {}", operation.operationId);
+
+                        operation.resultCompletableFuture.complete(
+                                Result.fail(Status.of(StatusCode.CANCELLED))
+                        );
+                    } else {
+                        logger.error("Fail cancel polling operation with id: {}", operation.operationId);
+                    }
+                }
+        );
+    }
+
+    @VisibleForTesting
+    static Status status(OperationProtos.Operation operation) {
+        StatusCode code = StatusCode.fromProto(operation.getStatus());
+        Double consumedRu = null;
+        if (operation.hasCostInfo()) {
+            consumedRu = operation.getCostInfo().getConsumedUnits();
+        }
+
+        return Status.of(
+                code,
+                consumedRu,
+                Issue.fromPb(operation.getIssuesList())
+        );
+    }
+
+    public static class Operation<Value extends Message> {
+
+        private final String operationId;
+        private final Class<Value> resultClass;
+        private final OperationManager operationManager;
+        private final CompletableFuture<Result<Value>> resultCompletableFuture;
+
+        public Operation(
+                String operationId,
+                Class<Value> resultClass,
+                OperationManager operationManager
+        ) {
+            this.operationId = operationId;
+            this.resultClass = resultClass;
+            this.operationManager = operationManager;
+            this.resultCompletableFuture = new CompletableFuture<>();
+        }
+
+        public Result<Value> getValue() {
+            return resultCompletableFuture.join();
+        }
+
+        public void cancel() {
```

Comment:
Maybe we can return CompletableFuture<Status> here?

Reply:
it's already done!)